File: examples/python/14_remote_access.py
"""Example 14 — Remote Access (Python)

Opens a tensogram file over HTTP using the remote backend.  A local
HTTP server with Range-request support is started in a background
thread so the example is fully self-contained.

NOTE: Requires building tensogram-python first:
    cd crates/tensogram-python && maturin develop
"""

from __future__ import annotations

import http.server
import threading

import numpy as np
import tensogram

# ── Minimal Range-capable HTTP server ─────────────────────────────────────────


def _make_range_handler(file_data: bytes):
    """Create an HTTP handler that serves *file_data* with Range support."""

    class Handler(http.server.BaseHTTPRequestHandler):
        def log_message(self, fmt, *args):
            pass

        def do_HEAD(self):
            self.send_response(200)
            self.send_header("Content-Length", str(len(file_data)))
            self.send_header("Accept-Ranges", "bytes")
            self.end_headers()

        def do_GET(self):
            range_header = self.headers.get("Range")
            if range_header and range_header.startswith("bytes="):
                spec = range_header[6:]
                if spec.startswith("-"):
                    n = int(spec[1:])
                    start = max(0, len(file_data) - n)
                    end = len(file_data)
                else:
                    parts = spec.split("-")
                    start = int(parts[0])
                    end = int(parts[1]) + 1 if parts[1] else len(file_data)
                    end = min(end, len(file_data))
                if start >= len(file_data) or start >= end:
                    self.send_response(416)
                    self.send_header("Content-Range", f"bytes */{len(file_data)}")
                    self.end_headers()
                    return
                chunk = file_data[start:end]
                self.send_response(206)
                self.send_header(
                    "Content-Range", f"bytes {start}-{end - 1}/{len(file_data)}"
                )
                self.send_header("Content-Length", str(len(chunk)))
                self.end_headers()
                self.wfile.write(chunk)
            else:
                self.send_response(200)
                self.send_header("Content-Length", str(len(file_data)))
                self.end_headers()
                self.wfile.write(file_data)

    return Handler


# ── Create a sample TGM file ─────────────────────────────────────────────────

meta = {
    "version": 2,
    "base": [
        {"mars": {"param": "2t", "step": 0, "date": "20260401"}},
        {"mars": {"param": "msl", "step": 0, "date": "20260401"}},
    ],
}
descriptors = [
    {
        "type": "ntensor",
        "shape": [72, 144],
        "dtype": "float32",
        "byte_order": "little",
        "encoding": "none",
        "filter": "none",
        "compression": "none",
    },
    {
        "type": "ntensor",
        "shape": [72, 144],
        "dtype": "float32",
        "byte_order": "little",
        "encoding": "none",
        "filter": "none",
        "compression": "none",
    },
]
rng = np.random.default_rng(42)
temperature = rng.standard_normal((72, 144)).astype(np.float32) + 293.15
pressure = rng.standard_normal((72, 144)).astype(np.float32) * 100 + 101325

tgm_bytes = tensogram.encode(
    meta, [(descriptors[0], temperature), (descriptors[1], pressure)]
)
print(f"Encoded {len(tgm_bytes)} bytes with 2 objects")

# ── Start local HTTP server ──────────────────────────────────────────────────

handler = _make_range_handler(tgm_bytes)
server = http.server.ThreadingHTTPServer(("127.0.0.1", 0), handler)
port = server.server_address[1]
url = f"http://127.0.0.1:{port}/forecast.tgm"
thread = threading.Thread(target=server.serve_forever, daemon=True)
thread.start()

print(f"Serving at {url}")

# ── URL detection ────────────────────────────────────────────────────────────

assert tensogram.is_remote_url(url)
assert not tensogram.is_remote_url("/tmp/local.tgm")
print(f"is_remote_url({url!r}) = True")

# ── Open remote file ─────────────────────────────────────────────────────────

f = tensogram.TensogramFile.open_remote(url, {})
print(f"\nOpened remote: source={f.source()}")
print(f"  is_remote = {f.is_remote()}")
print(f"  messages  = {f.message_count()}")

# ── Decode metadata (only fetches header, not full payload) ──────────────────

meta_obj = f.file_decode_metadata(0)
print(f"\nMetadata: version={meta_obj.version}")
print(f"  base[0] = {meta_obj.base[0]}")

# ── Decode descriptors (only fetches CBOR descriptor frames) ─────────────────

result = f.file_decode_descriptors(0)
descs = result["descriptors"]
print(f"\nDescriptors: {len(descs)} objects")
for i, d in enumerate(descs):
    print(f"  [{i}] shape={d.shape}  dtype={d.dtype}")

# ── Decode a single object (fetches only that object's frame) ────────────────

result = f.file_decode_object(0, 0)
arr = result["data"]
print(f"\nObject 0: shape={arr.shape}  dtype={arr.dtype}  mean={arr.mean():.2f}")
np.testing.assert_array_equal(arr, temperature)
print("  matches original temperature data")

result = f.file_decode_object(0, 1)
arr = result["data"]
print(f"Object 1: shape={arr.shape}  dtype={arr.dtype}  mean={arr.mean():.2f}")
np.testing.assert_array_equal(arr, pressure)
print("  matches original pressure data")

# ── Clean up ─────────────────────────────────────────────────────────────────

server.shutdown()
server.server_close()
thread.join(timeout=5)
print("\nRemote access example complete.")

File: examples/rust/src/bin/14_remote_access.rs
// (C) Copyright 2024- ECMWF and individual contributors.
//
// This software is licensed under the terms of the Apache Licence Version 2.0
// which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
// In applying this licence, ECMWF does not waive the privileges and immunities
// granted to it by virtue of its status as an intergovernmental organisation nor
// does it submit to any jurisdiction.

//! Example 14 — Remote Access
//!
//! Opens a tensogram file over HTTP using the remote backend.
//! A minimal HTTP server with Range-request support runs in a
//! background thread so the example is self-contained.
//!
//! Run with:  cargo run -p tensogram-rust-examples --features remote --bin 14_remote_access

use std::io::{Read, Write};
use std::net::TcpListener;
use std::sync::Arc;

use tensogram_core::{
    ByteOrder, DataObjectDescriptor, DecodeOptions, Dtype, EncodeOptions, GlobalMetadata,
    TensogramFile,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let meta = GlobalMetadata {
        version: 2,
        base: Vec::new(),
        ..Default::default()
    };
    let desc = DataObjectDescriptor {
        obj_type: "ntensor".to_string(),
        ndim: 2,
        shape: vec![72, 144],
        strides: vec![144, 1],
        dtype: Dtype::Float32,
        byte_order: ByteOrder::native(),
        encoding: "none".to_string(),
        filter: "none".to_string(),
        compression: "none".to_string(),
        params: std::collections::BTreeMap::new(),
        hash: None,
    };
    let data = vec![0u8; 72 * 144 * 4];
    let tgm_bytes =
        tensogram_core::encode::encode(&meta, &[(&desc, &data)], &EncodeOptions::default())?;
    println!("Encoded {} bytes", tgm_bytes.len());

    let tgm = Arc::new(tgm_bytes);
    let listener = TcpListener::bind("127.0.0.1:0")?;
    let port = listener.local_addr()?.port();
    let url = format!("http://127.0.0.1:{port}/forecast.tgm");

    let server_data = tgm.clone();
    std::thread::spawn(move || serve_http(listener, &server_data));

    println!("Serving at {url}");

    assert!(tensogram_core::remote::is_remote_url(&url));
    println!("is_remote_url = true");

    let file = TensogramFile::open_source(&url)?;
    println!(
        "\nOpened: source={} is_remote={} messages={}",
        file.source(),
        file.is_remote(),
        file.message_count()?,
    );

    let meta = file.decode_metadata(0)?;
    println!("Metadata: version={}", meta.version);

    let (_, descriptors) = file.decode_descriptors(0)?;
    println!("Descriptors: {} objects", descriptors.len());
    for (i, d) in descriptors.iter().enumerate() {
        println!("  [{i}] shape={:?} dtype={:?}", d.shape, d.dtype);
    }

    let (_, desc, decoded) = file.decode_object(0, 0, &DecodeOptions::default())?;
    println!(
        "\nObject 0: shape={:?} dtype={:?} bytes={}",
        desc.shape,
        desc.dtype,
        decoded.len()
    );
    assert_eq!(decoded, data);
    println!("  matches original data");

    println!("\nRemote access example complete.");
    Ok(())
}

fn serve_http(listener: TcpListener, data: &[u8]) {
    for stream in listener.incoming() {
        let mut stream = match stream {
            Ok(s) => s,
            Err(_) => continue,
        };
        let mut buf = vec![0u8; 4096];
        let n = match stream.read(&mut buf) {
            Ok(n) => n,
            Err(_) => continue,
        };
        let request = String::from_utf8_lossy(&buf[..n]);
        let is_head = request.starts_with("HEAD ");
        let range = parse_range(&request, data.len());

        match range {
            Some((start, end)) => {
                let chunk = &data[start..end];
                let header = format!(
                    "HTTP/1.1 206 Partial Content\r\n\
                     Content-Range: bytes {start}-{}/{}\r\n\
                     Content-Length: {}\r\n\
                     Accept-Ranges: bytes\r\n\
                     Connection: close\r\n\r\n",
                    end - 1,
                    data.len(),
                    chunk.len(),
                );
                let _ = stream.write_all(header.as_bytes());
                if !is_head {
                    let _ = stream.write_all(chunk);
                }
            }
            None => {
                let header = format!(
                    "HTTP/1.1 200 OK\r\n\
                     Content-Length: {}\r\n\
                     Accept-Ranges: bytes\r\n\
                     Connection: close\r\n\r\n",
                    data.len(),
                );
                let _ = stream.write_all(header.as_bytes());
                if !is_head {
                    let _ = stream.write_all(data);
                }
            }
        }
    }
}

fn parse_range(request: &str, file_len: usize) -> Option<(usize, usize)> {
    let range_line = request
        .lines()
        .find(|l| l.to_ascii_lowercase().starts_with("range:"))?;
    let spec = range_line.split("bytes=").nth(1)?.trim();
    if let Some(suffix) = spec.strip_prefix('-') {
        let n: usize = suffix.trim().parse().ok()?;
        Some((file_len.saturating_sub(n), file_len))
    } else {
        let mut parts = spec.split('-');
        let start: usize = parts.next()?.parse().ok()?;
        let end_str = parts.next()?;
        let end = if end_str.is_empty() {
            file_len
        } else {
            end_str.parse::<usize>().ok()? + 1
        };
        Some((start, end.min(file_len)))
    }
}
